HealthVault/Classes/Caching/Private/MHVThingCache.m

```diff
@@ -596,16 +684,261 @@ - (void)performSyncCompletionsWithSyncedCount:(NSInteger)syncedItemCount error:(
+                [MHVAsyncTask startSequenceOfTasks:methodsTasks withContinuationOption:MHVTaskContinueIfPreviousTaskWasNotCanceled];
+                
+                [MHVAsyncTask waitForAll:methodsTasks beforeBlock:^id(NSArray<MHVAsyncTaskResult *> *taskResults)
+                {
+                    for (MHVAsyncTaskResult *taskResult in taskResults)
+                    {
+                        if (taskResult.error)
+                        {
+                            cancel(taskResult);
+                            return nil;
+                        }
+                    }
+                    
+                    // Pass the database status to the next task
+                    finish(input);
+                    return nil;
+                }];
+            }
+        }];
+    }];
+}
+
+// Task to execute pending methods. Will CANCEL with an MHVAsyncTaskResult<NSError *> or FINISH with MHVAsyncTaskResult<nil>.
+- (MHVAsyncTask *)taskToExecutePendingMethod:(MHVPendingMethod *)pendingMethod
+{
+    return [[MHVAsyncTask alloc] initWithIndeterminateBlock:^(id input, void (^finish)(id), void (^cancel)(id))
+    {
+        [self.connection executeHttpServiceOperation:pendingMethod
+                                          completion:^(MHVServiceResponse * _Nullable response, NSError * _Nullable error)
+        {
+            if (error)
+            {
+                cancel([MHVAsyncTaskResult withError:error]);
+            }
+            else
+            {
+                [self.database deletePendingMethods:@[pendingMethod]
+                                         completion:^(NSError * _Nullable error)
+                {
+                    if (error)
+                    {
+                        cancel([MHVAsyncTaskResult withError:error]);
+                    }
+                    else
+                    {
+                        finish([MHVAsyncTaskResult withResult:nil]);
+                    }
+                }];
+            }
+        }];
+    }];
+}
+
+// Task to get the record operations since the last sync. Will CANCEL with an MHVAsyncTaskResult<NSError *> or FINISH with MHVAsyncTaskResult<NSNumber *>.
+- (MHVAsyncTask *)taskForRecordOperationsWithRecordId:(NSString *)recordId
+{
+    return [[MHVAsyncTask alloc] initWithIndeterminateBlock:^(MHVAsyncTaskResult *input, void (^finish)(id), void (^cancel)(id))
+    {
+        id<MHVCacheStatusProtocol> status = input.result;
+        
+        if (!status)
+        {
+            cancel([MHVAsyncTaskResult withError:[NSError MHVCacheError:@"Could not determine the status of the cache database."]]);
+            return;
+        }
+        
+        // If the lastCacheConsistencyDate value is set and the time between now and the lastCompletedSyncDate is < syncIntervalSeconds don't sync.
+        if (status.lastCacheConsistencyDate && fabs([status.lastCompletedSyncDate timeIntervalSinceNow]) < self.cacheConfiguration.syncIntervalSeconds)
```

Comment:
I wonder if we should have a Force Sync option?  An app could have their own push notifications, so know something changed and should be synced.